components/layout_2020/display_list/stacking_context.rs

```diff
@@ -1635,17 +1636,38 @@ impl BoxFragment {
     /// Returns the 4D matrix representing this fragment's transform.
     pub fn calculate_transform_matrix(&self, border_rect: &Rect<Au>) -> Option<LayoutTransform> {
         let list = &self.style.get_box().transform;
+        // https://drafts.csswg.org/css-transforms-2/#individual-transforms
+        let rotate = match self.style.clone_rotate() {
+            GenericRotate::Rotate(angle) => (0., 0., 1., angle),
+            GenericRotate::Rotate3D(x, y, z, angle) => (x, y, z, angle),
+            GenericRotate::None => (0., 0., 1., Angle::zero()),
+        };
+        let scale = match self.style.clone_scale() {
+            GenericScale::Scale(sx, sy, sz) => (sx, sy, sz),
+            GenericScale::None => (1., 1., 1.),
+        };
+        let translation = match self.style.clone_translate() {
+            GenericTranslate::Translate(x, y, z) => LayoutTransform::translation(
+                x.to_used_value(border_rect.size.width).to_f32_px(),
+                y.to_used_value(border_rect.size.height).to_f32_px(),
```

Comment:
I think these should resolve percentages in the same way as the transform, so I would so the `let border_rect = au_rect_to_length_rect(border_rect)` first and then

```suggestion
                x.resolve(border_rect.size.width).px(),
                y.resolve(border_rect.size.height).px(),
```